### ISSUE TITLE
Core: forms/inputs

### ISSUE THREAD

**Issue by coryrylan** [vmware/clarity#3996]
Build base input wrapper to support native html inputs.
- [x] Basic a11y (auto id/for assignment)
- [x] Required labels, if label is not provided log error/warning
- [x] Error state support
- [x] layout property providing the layout types (`horizontal`, `vertical`)
- [x] icon groups in text based inputs, both auto apply and custom
- [x] Stateless and rely on slotted native inputs for best framework support, do not manage validation
- [x] Inputs must leverage native html inputs for frameworks to hook into. All inputs must be projected from light DOM and cannot be split across boundary between label and input for a11y.
- [x] groups example screenshot [here](https://github.com/vmware/clarity/pull/4403#pullrequestreview-396876833)

## Form Group

The form group sets the layout of a group of cds-controls.

```html
<cds-form-group layout="horizontal | vertical | horizontal-inline | vertical-inline | compact">
  <cds-input>
    <label>Text</label>
    <input type="text" placeholder="name" />
  </cds-input>

  <cds-input>
    <label>Text</label>
    <input type="text" placeholder="name" />
  </cds-input>
</cds-form-group>
```

## Control

Applies basic layout styles. For custom controls provided by the application.
All input controls will inherit layout and status support.

```html
<cds-control status="error|default">
  <label>label text</label>
  <app-input></app-input>
  <cds-control-message status="neutral | error | success">
    message text
  </cds-control-message>
</cds-control>
```

## Control Group

The control group sets the layout of a group of cds-controls. This is used for
related controls such as a radio list or checkbox list.

```html
<cds-control-group layout="horizontal | vertical | horizontal-inline | vertical-inline | compact">
  <cds-control>
    <label>Option 1</label>
    <some-input></some-input>
  </cds-control>

  <cds-control>
    <label>Option 2</label>
    <some-input></some-input>
  </cds-control>
</cds-control-group>
```

### Input 📝

Most generic html text based input types can work with cds-input.

```html
<cds-input>
  <label>Text</label>
  <input type="text" placeholder="name" />
</cds-input>

<cds-input>
  <label>Email</label>
  <input type="email" placeholder="test@test.com" />
</cds-input>

<cds-input>
  <label>Number</label>
  <input type="number" placeholder="age" />
</cds-input>
```

### Textarea 📝

```html
<cds-textarea>
  <label>Description</label>
  <textarea></textarea>
</cds-textarea>
```

### Select 🗄️

```html
<cds-select>
  <label>Select an Option</label>
  <select>
    <option>Option One</option>
    <option>Option Two</option>
    <option>Option Three</option>
  </select>
</cds-select>
```

### Checkbox ✔️

```html
<cds-checkbox>
  <input type="checkbox">
  <label>Remember Me</label>
</cds-checkbox>

<cds-checkbox-group>
  <legend>Select at least one</legend>
  <cds-checkbox>
    <label>Check One</label>
    <input type="checkbox" checked>
  </cds-checkbox>

  <cds-checkbox>
    <label>Check Two</label>
    <input type="checkbox">
  </cds-checkbox>
</cds-checkbox-group>
```

### Radio 🔘

```html
<cds-radio-group name="region">
  <legend>Choose a Country</legend>
  <cds-radio>
    <label>North America</label>
    <input type="radio" value="north-america" />
  </cds-radio>
  <cds-radio>
    <label>South America</label>
    <input type="radio" value="south-america" />
  </cds-radio>
  <cds-radio>
    <label>Europe</label>
    <input type="radio" value="europe" disabled />
  </cds-radio>
</cds-radio-group>
```

### Datalist 🗄️

```html
<cds-datalist>
  <label>label text</label>
  <input placeholder="placeholder text" />
  <datalist>
    <option value="Item 1"></option>
    <option value="Item 2"></option>
    <option value="Item 3"></option>
  </datalist>
</cds-datalist>
```

### Password 🔒

```html
<cds-password>
  <label>New Password</label>
  <input type="password" />
</cds-password>
```

### Time 🕐

```html
<cds-time>
  <label>Select Time</label>
  <input type="time" min="09:00" max="18:00" />
</cds-time>
```

### Input Group 📝

Groups can allow multiple input types as well as actionable/readonly control actions. Control actions can be prefix/suffix labels or icon buttons. This will allow all input types use the same system for icon buttons.

```html
<cds-input-group>
  <legend>Billing Limit</legend>
  <cds-input>
    <label>Billing Limit</label>
    <input placeholder="100.00" type="number" value="1987.56" />
  <cds-control-action action="prefix" readonly>$</cds-control-action>
  </cds-input>
  <cds-select cds-layout="align:shrink">
    <label>Currency</label>
    <select>
      <option>USD</option>
      <option>EUR</option>
    </select>
  </cds-select>
  <cds-control-message>Maximum host budget for billing period</cds-control-message>
</cds-input-group>
```

### Toggle ✔️

```html
<cds-toggle>
  <label>Toggle</label>
  <input type="checkbox" checked />
</cds-toggle>
```

### Range 🎚️

```html
<cds-range>
  <label>Select Range</label>
  <input type="range" min="0" max="100" value="90" step="10" />
</cds-range>
```

### Datepicker 📅

Single date datepicker leverages native input. Visually will show custom picker but set the underlying input. Min and max is controlled by reading the min/max set on the native control.
The value is always `YYYY-MM-DD` following the native ISO8601 spec https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date 

```html
<cds-date>
  <label>Select Date</label>
  <input type="date"  value="2018-07-22" min="2018-01-01" max="2018-12-31"/>
</cds-date>
```

Range datepicker can be enabled by providing a second native date input. Visually neither is shown but our custom datepicker is rendered. It will set the first date in range to the first input and last date to the second. Min and max range can be determined by the min/max set on the two native inputs.

```html
<cds-date>
  <label>Start Date</label>
  <input type="date"  min="2018-01-01" />

  <label>End Date</label>
  <input type="date"  max="2018-12-31" />
</cds-date>
```

### Rating ⭐⭐⭐⭐⭐

Star rating style component. Follows same pattern with leveraging a native control to determine, step and start count.

```html
<cds-rating>
  <label>Item Review</label>
  <input type="number" min="0" max="5" sep="0.5" />
</cds-rating>
```

**Comment by gnomeontherun** [comment]
Do we need to create a custom container for each control like `cds-input`? I suppose this is is to handle the potential quirks of the layout, but is it necessary?

Checkbox groups might be validated as one unit, or as individual checkboxes that are just in a 'group'. For example, you might have a list of 5 where you must select 3 items, which would be validating the group as one unit. You might also have a set of checkboxes where selecting two items might be invalid and you'd need to indicate specifically which one is invalid. Maybe we don't do this (we don't really have this now) but wanted to raise it.

Should add the time picker.

How do you modify the layouts if you want to put them into a grid, or change the width of the label?

Are inputs/selects going to 'fill' their containers? 

What about input groups API?

No compact layout?

Do you do anything to wire up the label/input for/id attributes?

**Comment by coryrylan** [comment]
1. No you will not need a custom container for each text based input type. Only when the input diverges from a text based input (range, switch) or the input has full custom implementation (datepicker, range).

2. Checkbox group is only for layout/style. We can allow an error message to be rendered outside the control wrapper but within a control group to show errors tied to a group. That should allow any combination of checkboxes to be validated against each other.

3. Input type "time" would fall under the `cds-input` until we implement a custom UI on top of that like we do with datepicker. Timepicker will be restricted to the native time APIs. I'll add it to the advanced controls list. All of the non custom text based inputs will have a demo in storybook as well.

4. Layouts are still a wip and I am experimenting with this to find the right balance in flexibility and convenience.

5. I would prefer to fill by default but fill is undecided. My prototypes currently dont like clr-ui but this would be the time to either make it the default. Regardless it should provide a way for them to fill or fit. 

6. Input groups still needs work as well as some more investigation on how to support it with a11y and ensuring there are no invalid states. This may need to be post mvp but I'll work up a API/prototype to document.

7. Compact layout will be managed with the layout API and is still a wip as well.

8. label/input and id/aria associations will be wired up automatically similar to clr-angular (this is working in the prototype)
